### PR TITLE
[fix, test] Fix recall metric bug, add test

### PR DESCRIPTION
**Pull request by brettallenyo**
- Fix infinite recursion error #725
- Add back mistakenly removed code
- Add test for recall metric at 1, 5, and 10

This is a response to bug #725 

I believe `process_ranks` may have been accidentally deleted in this commit so I added it back: 
https://github.com/facebookresearch/mmf/commit/6b6f4fc221c46f8c55d55a06e110f3d8f9ca2417

I also added a testing file, which basically ranks all of the expected values in increasing order (so the last candidate is always the chosen one).

Both the expected values and predicted values are (10, 100) tensors. As the index of top level array increases, I alter the final value to be decreasing by 2, so the chosen candidates are 198 (maximum value 99*2), 191, 189, 187, ...

I decrease it in this way so that each of the candidates become out of the bounds for the increasing Recall Metrics. The rankings are `tensor([ 1,  4,  5,  6,  7,  8,  9, 10, 11, 12])`

This makes the results varying for each metric:

`RecallAt1` --> 1/10 = .1
`RecallAt5` --> 3/10 = .3
`RecallAt10` --> 8/10 = .8

I would love for people to take a look at `metrics.py` line `462` (`process_ranks`)
There are print statements in here from the previous commit, but I do not see print statements anywhere else in the modules, should I delete these, leave them, do we have a different logging style? Thanks!

### CODE REVIEW

File: mmf/modules/metrics.py [hunk 1]
```diff
@@ -459,6 +459,22 @@ def get_gt_ranks(self, ranks, ans_ind):
             gt_ranks[i] = int(ranks[i, ans_ind[i].long()])
         return gt_ranks
 
+    def process_ranks(self, ranks):
+        num_opts = 100
+
+        # none of the values should be 0, there is gt in options
+        if torch.sum(ranks.le(0)) > 0:
+            num_zero = torch.sum(ranks.le(0))
+            print(f"Warning: some of ranks are zero: {num_zero}")
```
Comment (ytsheng) [inline mmf/modules/metrics.py:468]: let's try to use the warning module, and not have print statements.

**Review by ytsheng — COMMENTED** [review]
Thanks for adding tests! I have some small nits, but otherwise it looks good!

File: mmf/modules/metrics.py (hunk 1 above)
Comment (brettallenyo) [reply mmf/modules/metrics.py]: That's a great idea, just changed it :) thanks!

**Review by vedanuj — APPROVED** [review]
This looks great! Can you import to Phabricator?

File: mmf/modules/metrics.py [hunk 2]
```diff
@@ -459,6 +460,22 @@ def get_gt_ranks(self, ranks, ans_ind):
             gt_ranks[i] = int(ranks[i, ans_ind[i].long()])
         return gt_ranks
 
+    def process_ranks(self, ranks):
+        num_opts = 100
+
+        # none of the values should be 0, there is gt in options
+        if torch.sum(ranks.le(0)) > 0:
+            num_zero = torch.sum(ranks.le(0))
+            warnings.warn(f"Warning: some of ranks are zero: {num_zero}")
```
Comment (apsdehal) [inline mmf/modules/metrics.py:469]: Don't need to mention "Warning:", that will be added automatically by warning module

File: mmf/modules/metrics.py [hunk 3]
```diff
@@ -459,6 +460,22 @@ def get_gt_ranks(self, ranks, ans_ind):
...
+            ranks = ranks[ranks.gt(0)]
+
+        # rank should not exceed the number of options
+        if torch.sum(ranks.ge(num_opts + 1)) > 0:
+            num_ge = torch.sum(ranks.ge(num_opts + 1))
+            warnings.warn(f"Warning: some of ranks > 100: {num_ge}")
```
Comment (apsdehal) [inline mmf/modules/metrics.py:475]: Same here.

**Review by apsdehal — APPROVED** [review]
Looks good! Thanks for fixing this.
